messaging/src/main/java/org/axonframework/eventhandling/TrackingEventProcessor.java

```diff
@@ -210,16 +287,50 @@ private void releaseToken(Segment segment) {
         }
     }
 
+    /**
+     * Indicates whether the {@code eventMessage} identified with given {@code token} should be processed as part of the
+     * given {@code segment}. This implementation is away of merge tokens and will recursively detect the (sub)segment
+     * in which an event should be handled.
+     *
+     * @param token   The token to check segment validity for
+     * @param segment The segment to process the event in
+     * @return {@code true} if this event should be handled, otherwise {@code false}
+     */
+    protected List<Segment> processingSegments(TrackingToken token, Segment segment) {
+        Optional<MergedTrackingToken> mergedToken = WrappedToken.unwrap(token, MergedTrackingToken.class);
+        if (mergedToken.isPresent()) {
+            Segment[] splitSegments = segment.split();
+            List<Segment> segments = new ArrayList<>();
+            if (mergedToken.get().isLowerSegmentAdvanced()) {
+                segments.addAll(processingSegments(mergedToken.get().lowerSegmentToken(), splitSegments[0]));
+            }
+
+            if (mergedToken.get().isUpperSegmentAdvanced()) {
+                segments.addAll(processingSegments(mergedToken.get().upperSegmentToken(), splitSegments[1]));
+            }
+            return segments;
+        }
+        return singletonList(segment);
+    }
+
     private void processBatch(Segment segment, BlockingStream<TrackedEventMessage<?>> eventStream) throws Exception {
         List<TrackedEventMessage<?>> batch = new ArrayList<>();
         try {
             checkSegmentCaughtUp(segment, eventStream);
-            TrackingToken lastToken = null;
+            TrackingToken lastToken;
+            List<Segment> processingSegments;
             if (eventStream.hasNextAvailable(1, SECONDS)) {
-                for (int i = 0; i < batchSize * 10 && batch.size() < batchSize && eventStream.hasNextAvailable(); i++) {
+                final TrackedEventMessage<?> firstMessage = eventStream.nextAvailable();
+                lastToken = firstMessage.trackingToken();
+                processingSegments = processingSegments(lastToken, segment);
+                if (canHandle(firstMessage, processingSegments)) {
+                    batch.add(firstMessage);
+                }
+                for (int i = 0; singletonList(segment).equals(processingSegments) && i < batchSize * 10 && batch.size() < batchSize
```

Comment:
This condition looks massive. Is it possible to extract it in a method with a meaningful name?